great3sims/psf.py

```diff
@@ -1393,64 +1405,64 @@ def generateFieldParameters(self, rng, field_index):
-        # draw_psf_src is named foo.fits, the psfs are named foo_n.fits.
-       
-        index = self.draw_psf_src.find('.fits')
-        if index < 0: format = self.draw_psf_src + '_%d'
-        else:
-            format = self.draw_psf_src[:index] + '_%d' + self.draw_psf_src[index:]
-
-        # if the file self.draw_psf_src exists, assume it is a multi-hdu
-        # file containing a collection of psfs, and open if for access
-        hdus = None
-        if os.path.exists(self.draw_psf_src):
-            hdus = pyfits.open(self.draw_psf_src)
-            psf_count = len(hdus)
-
-        # if not, just enumerate the individual numbered fits files in
-        # the same directory
-        else:
-            psf_count = 0
-            while os.path.exists(format%(psf_count+1,)):
-                psf_count += 1
-            print "psf_count = ", psf_count
-        # now randomly assign availabel psfs to the individual records,
-        # expanding those which are not already there from the hdus
+        #   The psfInfos structure is a list of all the available Psfs,
+        #   Randomly assign available psfs to the individual records, and mark each
+        #   one which is choosen for expansion.  Note that the "psf_group" number is
+        #   another way to refer to the file named psfInfo[0], but using a number.
         for record in catalog:
             # select a random hdu
-            psf_number = int(math.floor(rng() * psf_count)) + 1 
-            record["psf_number"] = psf_number
-            filename = format%psf_number
-            if not os.path.exists(filename):
-                psf = pyfits.PrimaryHDU(hdus[psf_number-1].data)
+            psfInfo = self.psfInfos[int(math.floor(rng() * len(self.psfInfos)))]
+            record["psf_group"] = psfInfo[1]
+            record["psf_index"] = psfInfo[2]
+            record["psf_number"] = psfInfo[3]
+            psfInfo[4] = True
+        # write the tagged entries in the order of the psfInfos file.  This
+        # is sorted by library, minimizing the amount of file I/O required to fetch the psfs
+        dirname = os.path.dirname(self.draw_psf_src)
+        currentPsfFile = None
+        for i in range(len(self.psfInfos)):
+            psfInfo = self.psfInfos[i]
+            if psfInfo[4]:
+                if psfInfo[0] != currentPsfFile:
+                    if not currentPsfFile is None:
+                        hdus.close()
+                        del hdus
+                    currentPsfFile = psfInfo[0]
+                    filename = dirname + "/" + psfInfo[0]
+                    print "opening ", filename
+                    hdus = pyfits.open(filename)
+                print len(hdus), psfInfo[2]
+                psf = pyfits.PrimaryHDU(hdus[psfInfo[2]-1].data)
                 psf.header.append("GS_SCALE")
-                psf.header["GS_SCALE"] = 0.02
-                psf.writeto(filename)
-        if hdus: hdus.close()
+                psf.header["GS_SCALE"] = 0.2
```

Comment:
We wrote a ``GS_SCALE`` parameter in ``makePsfLibrary.py``. Shouldn't we just be copying that over instead of redefining it? What happens if you later decide to change it, but forget that it's defined in (at least) two places.

Again, I think a comment describing what ``GS_SCALE`` is and why you have chosen 0.2 would be appropriate (I note it was originally 0.02 in the above, which makes me think this is doubly necessary).